### PR TITLE
[cudnn] Fix cudnn dlls

### PR DESCRIPTION
**Pull request by Olli1080**
This fixes that dlls won't be installed under windows when using TARGET_RUNTIME_DLLS generator expression in cmake.
The cause is a misspelled variable. 

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [ ] ~~SHA512s are updated for each updated download.~~
- [ ] ~~The "supports" clause reflects platforms that may be fixed by this new version.~~
- [ ] ~~Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.~~
- [ ] ~~Any patches that are no longer applied are deleted from the port's directory.~~
- [ ] ~~When updating the upstream version, the `"port-version"` is reset (removed from `vcpkg.json`).~~
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Comment by Olli1080** [comment]
I've found another edge case.
When copying cudnn into the cuda installation the path for the dlls is located at ../bin relative to the includes.
But when using the installer the include and bin directory are augmented by a folder with a cuda version. (e.g. include/12.6/; bin/12.6/

File: ports/cudnn/FindCUDNN.cmake [hunk 1]
```diff
@@ -86,7 +86,14 @@ find_package_handle_standard_args(CUDNN
...
   list(TRANSFORM CUDNN_DLL_DIR APPEND "/../bin")
-  find_file(CUDNN_LIBRARY_DLL NAMES cudnn64_${CUDNN_VERSION_MAJOR}.dll PATHS ${CUDNN_DLL_DIR})
+  find_file(CUDNN_LIBRARY_DLL NAMES cudnn64_${_CUDNN_VERSION_MAJOR}.dll PATHS ${CUDNN_DLL_DIR})
+
+  if (NOT CUDNN_LIBRARY_DLL)
+    get_filename_component(_CUDA_VERSION ${CUDNN_INCLUDE_DIR} NAME)
+    set(CUDNN_DLL_DIR ${CUDNN_INCLUDE_DIR})
+    list(TRANSFORM CUDNN_DLL_DIR APPEND "/../../bin/${_CUDA_VERSION}")
+    find_file(CUDNN_LIBRARY_DLL NAMES cudnn64_${_CUDNN_VERSION_MAJOR}.dll PATHS ${CUDNN_DLL_DIR})
+  endif()
```
Comment (BillyONeal) [inline ports/cudnn/FindCUDNN.cmake:96]: It seems like this would be better done by just adding the path in question to PATHS in the first `find_file` rather than doing separate `find_file`s like this?

Comment (Olli1080) [reply ports/cudnn/FindCUDNN.cmake:96]: Yes it certainly does. I've also found some other issues which i'll push a fix for soon.

**Review by FrankXie05 — APPROVED** [review]

**Comment by JavierMatosD** [comment]
Tagging this team review on account that we don't allow find modules in the repo. I know this is already the status quo but given that this is the ports 15th revision, these are substantial changes to the find module, and `FindCUDA.cmake` is deprecated, I'm not exactly sure how to proceed here.

**Comment by BillyONeal** [comment]
@Olli1080 To translate what @JavierMatosD into something you can help with: Normally, find modules must not be installed, and that this port ever did that is kind of an issue.

>The cause is a misspelled variable.

Can you fix just the misspelled variable rather than trying to do a full overhaul like this? Because as written it argues that you should go all the way and fix it being a config in the first place.

It's also not clear what this is supposed to be doing in the first place considering there doesn't appear to be standard target names for cudnn.

**Comment by Olli1080** [comment]
@BillyONeal thank you for the information.

The issue with the misspelled variable was using CUDNN_VERSION_MAJOR instead of _CUDNN_VERSION_MAJOR (missing _ in front of name). This only solved finding cudnn64_${_CUDNN_VERSION_MAJOR}.dll which is only usefull for cudnn 7.6.5 which is the lowest supported version (according to MINIMUM_CUDNN_VERSION). After version 7.6.5 cudnn was split into multiple libraries (and dlls) which would make add_library neccessary for every library needed.

Could you please explain what the alternative is to using a find module?

The changes are supposed to allow to automatically pull in the neccessary runtime dependencies with $<TARGET_RUNTIME_DLLS:tgt> based on the used cudnn version and chosen dependencies.
I neither found any standardized module names only the inter library usage/link dependencies.

I can understand if the changes are to big to justify supporting $<TARGET_RUNTIME_DLLS:tgt> but it would still be a nice to have.

I'd be happy to help if any changes are neccessary.

**Comment by BillyONeal** [comment]
Sorry I edited your comment above to add backticks (\`) so that GitHub would format it correctly.

> The issue with the misspelled variable was using `CUDNN_VERSION_MAJOR` instead of `_CUDNN_VERSION_MAJOR` (missing `_` in front of name).

I think we would be happy to take a change that fixes this typo without effectively rewriting these find modules.

> Could you please explain what the alternative is to using a find module?

Normally find modules belong only in "consuming" projects. Someone trying to do publishing like this are normally expected to use configs rather than find modules. See 'module mode' vs 'config mode' here: https://cmake.org/cmake/help/latest/command/find_package.html#search-modes

> The changes are supposed to allow to automatically pull in the necessary runtime dependencies with `$<TARGET_RUNTIME_DLLS:tgt>` based on the used cudnn version and chosen dependencies.
> I neither found any standardized module names only the inter library usage/link dependencies.

If there's no existing standardized target name for this, then that means there aren't any existing users who expect `$<TARGET_RUNTIME_DLLS:tgt>` to be associated with these particular target names. As a result, I'm not sure that's a can of worms we really want to open given that we are merely referencing a 'system installed' copy of these libraries and not distributing them ourselves. Put another way, I expect that there will be bugs in this due to different ways different systems deploy these libraries, or due to legitimate differences between versions of cudnn. Then, people will reasonably feel like the vcpkg maintainers are on the hook to resolve differences, which we are ill equipped to do being not cudnn users ourselves. It is truly unfortunate that cudnn's licensing restricts publishing their bits like anything else in our registry :( .

**Comment by Olli1080** [comment]
Thank you for the clarifications.
It's a pity that cudnn won't work with this cmake feature but i understand where you are coming from.

Since it is not intended to maintain the copy process of the dlls it may be better to just remove the section for `CUDNN_LIBRARY_DLL` since it may cause more confusion in the future.
`CUDNN_LIBRARY_DLL` is only useful (in the scope of the existing find module) if used in conjunction `$<TARGET_RUNTIME_DLLS:tgt>`. This is also the reason why i started to introduce bigger changes to the module in the first place since i assumed that it was intentionally supported but forgotten about.

TLDR: It might be better to just remove the corresponding lines instead, to avoid future confusion.

If you are okay with these changes, I will create a new pull request for them.

**Comment by BillyONeal** [comment]
> TLDR: It might be better to just remove the corresponding lines instead, to avoid future confusion.
> 
> If you are okay with these changes, I will create a new pull request for them.

If there is nobody looking for these as far as we can tell I would be happy for the variables to be removed. I'm not sure how to prove or disprove that though.

**Comment by vicroms** [comment]
We have [closed this PR due to inactivity](https://learn.microsoft.com/vcpkg/contributing/maintainer-guide#close-inactive-prs). Please create a new PR if you're still interested in getting these changes merged.
